### DISCUSSION TITLE
okular: cannot print to pdf

### DISCUSSION THREAD

**Discussion by gcb** [netblue30/firejail#6402] [Q&A] [answered]
My system do not have cups, only cupslibs.

most applications (outside of firejail) will show a "print to PDF" option as a "printer".

Same application with default git profiles will not show such printer. Shows an empty printer list.

Not sure what to allow on applications such as okular to enable that PDF printer. I only see alternative-etc for the non-existing dir /etc/cups. Not sure what else to enable. or even what is providing that printer in my system.

**Comment by kmk3** [comment]
Basic debugging information is missing; please follow the bug report template:

* <https://github.com/netblue30/firejail/issues/new?template=bug_report.md>

Also, if you run the following command:

```sh
firejail --trace=trace.txt /usr/bin/okular $file
```

And try to print the file, then close okular, what is the content of trace.txt?

**Reply by gcb** [reply]
### Description
^

### Steps to Reproduce

1. Run in bash `LC_ALL=C firejail okular` (`LC_ALL=C` to get a consistent
   output in English that can be understood by everybody)
2. "Ctrl+P"

### Expected behavior

See the QT default "print to PDF" pseudo-printer.

### Actual behavior

there's no printers in the printer drop down.

### Behavior without a profile

I see the default QT pseudo printer "Print to PDF"

### Additional context

The necessary libs are whitelisted OK on the default profile available here (`okular.profile`). But the dbus messages are blocked.

### Environment

```
firejail version 0.9.72

Compile time support:
        - always force nonewprivs support is disabled
        - AppArmor support is enabled
        - AppImage support is enabled
        - chroot support is enabled
        - D-BUS proxy support is enabled
        - file transfer support is enabled
        - firetunnel support is disabled
        - IDS support is disabled
        - networking support is enabled
        - output logging is enabled
        - overlayfs support is disabled
        - private-home support is enabled
        - private-cache and tmpfs as user enabled
        - SELinux support is disabled
        - user namespace support is enabled
        - X11 sandboxing support is enabled
Linux 6.9.9-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 12 Jul 2024 00:06:53 +0000 x86_64 GNU/Linux
```


### Checklist



- [x] The issues is caused by firejail (i.e. running the program by path (e.g. `/usr/bin/vlc`) "fixes" it).
- [x] I can reproduce the issue without custom modifications (e.g. globals.local).
- [x] The program has a profile. (If not, request one in `https://github.com/netblue30/firejail/issues/1139`)
- [x] The profile (and redirect profile if exists) hasn't already been fixed [upstream](https://github.com/netblue30/firejail/tree/master/etc).
- [x] I have performed a short search for similar issues (to avoid opening a duplicate).

### Log

No logs provide any information for dbus access.

**Comment by ghost** [comment]
You could try to comment options in okular.profile and try to pinpoint what's blocking that "print to PDF" option. That's always a bit time-consuming obviously. I'm not at all familiar with KDE, and I don't know what provides such a PDF printer option either. Based on the profile my first assumptions are we're either missing a command (in private-bin) and/or extra cups-related files from /usr/share are needed. Maybe the below can jump-start the search for the blocking/missing option(s):

```sh
$ cat ~/.config/firejail/okular.local
whitelist /usr/share/cups
whitelist /usr/share/ppd

ignore private-bin
ignore private-etc
```

Once we've established what's needed, we can try to undo the effects of loosening the sandbox with the ignore private-{bin,etc} suggestions.

**Reply by gcb** [reply]
> cups, ppd

my system have neither. So i don't think they are involved in [QT's print to pdf](https://doc.qt.io/qt-6/qtprintsupport-index.html) functionality.

⚠️ edit: my bad, this info bellow is all wrong.
~~~
creating a `.local` file with:

```
dbus-user filter
```
does enables printing. But that is hardly a restrictive solution and opens up much more than just printing.

Would be nice to figure out a minimum-safe and have a common-kde-qtprint sort of file.
~~~

I think i had messed up with the .profile, and then wasn't running `firecfg` when changing the .local one :(

**Reply by ghost** [reply]
The default okular.profile doesn't block user/system dbus by default. You could try to narrow things down with a tool like [QDbusViewer](https://doc.qt.io/qt-5/qdbusviewer.html) and add the additional `dbus-user.own` / `dbus-user.talk` rules. Also, can you still get printing when you enable `dbus-system none`?

**Reply by gcb** [reply]
oh, indeed. the default profile, nor the one shipped in my distro, blocks dbus! yet clearing it on a .local file fixed the issue :/

i will have to learn how to debug profiles i guess. something in the commons file might be screwing up with dbus.

...wondering if not setting any dbus options, but setting `net none` breaks dbus?

**Reply by rusty-snake** [reply]
> but setting net none breaks dbus?

Only if dbus uses an abstract socket. i.e. `$DBUS_SESSION_BUS_ADDRESS` starts with `unix:abstract` rather than `unix:path`.

Abstract sockets are not recommended because you can not block them (e.g. `dbus-user none` does not work).

**Reply by gcb** [reply]
thanks. So not the case: `DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus`

**Reply by gcb** [reply]
i've edited the comment at the start of this thread. It was not dbus. I was doing something wrong between tests.

I still have no idea what enables printing...

**Reply by gcb** [reply]
`ignore private-bin` alone in the .local file does fix printing!

running `firejail --debug okular` with and witouth it produces almost the same output. the only difference is

without `ignore private-bin` (and no printing):
```
Disable /usr/local/sbin
Warning (blacklisting): cannot open /usr/local/sbin/at: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/busybox: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/chage: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/chfn: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/chsh: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/crontab: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/doas: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/evtest: Permission denied
```

with `ignore private-bin` (and printing OK):
```
Disable /usr/local/sbin
Warning (blacklisting): cannot open /usr/local/sbin/at: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/busybox: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/chage: Permission denied
Disable /usr/bin/chage
Warning (blacklisting): cannot open /usr/local/sbin/chfn: Permission denied
Disable /usr/bin/chfn
Warning (blacklisting): cannot open /usr/local/sbin/chsh: Permission denied
Disable /usr/bin/chsh
Warning (blacklisting): cannot open /usr/local/sbin/crontab: Permission denied
Disable /usr/bin/crontab
Warning (blacklisting): cannot open /usr/local/sbin/doas: Permission denied
Disable /usr/bin/doas
Warning (blacklisting): cannot open /usr/local/sbin/evtest: Permission denied
Warning (blacklisting): cannot open /usr/local/sbin/expiry: Permission denied
Disable /usr/bin/expiry
```
(it shows those extra "Disable ..." lines)

**Reply by gcb** [reply]
i don't see any fork or the likes on the source https://github.com/qt/qtbase/blob/ff1ba4f1c9da383b357745c2261de5b0050cdf3c/src/printsupport/kernel/qprintengine_pdf.cpp

**Reply by ghost** [reply]
> ignore private-bin alone in the .local file does fix printing!

Interesting. So we need to find which executable(s) are missing from the original private-bin for supporting printing in KDE. As requested by @kmk3 [above](https://github.com/netblue30/firejail/discussions/6402#discussioncomment-10016309) you can run:

> firejail --trace=trace.txt /usr/bin/okular $file
And try to print the file, then close okular, what is the content of trace.txt?

If you can provide that `trace.txt` here that would help us find what we're looking for.

Now I'm wondering if okular (and possibly other KDE apps) needs a shell for printing support... Easily tested with the below:

```sh
$ cat ~/.config/firejail/okular.local
include allow-bin-sh.inc
private-bin bash,sh
```

**Comment by gcb** [comment]
The trace was more helpful than debug (apologies, at first I assumed it was just an option to save the same infro from `--debug` to a file)

So, besides learning okular calls stat on *all* my home files over and over for no good reason, and that it reads the mime files some hundred times each... the trace showed it needs access to ps2pdf on linux.

this is the file on my .local file now (from my tests I believe it *ADD*s to the same value in okular.profile)
```
private-bin ps2pdf
```
...I guess we can add this on a common-kde.inc file? Is there such file? Anything that prints in KDE will required it for the "to PDF" functionality. But it's also a exploit gateway as it deals with two of the worst file formats :)

the `okular.profile` one have things which I believe are very out of date
* `kbuildsycoca4` is several years out of use. I mean even by 'debian-stable' standards that is deprecated. kde5 didn't have this as far as I can tell, and 6 (current) reintroduced as kbuildsycoca6. but grepping for `kbuildsycoca` in my traces showed nothing. I think it can go.
* `okular` fails with "Error: no suitable /usr/bin/okular executable found" not sure if okular needs to call itself? how does that work, having the same bin of the masked program in `private-bin`? does it call the unmasked one? 
* `kdeinit4` deprecated like the first one above. not reintroduced in 6.
* `lpr` i couldn't find KDE or QT using it nowadays. I think that is from 4- times. 5 and 6 only use the native method (which i learned here that depends on `ps2pdf` and have support for `cups`)
* `unar`,`unrar` I don't have either so :shrug:

**Reply by ghost** [reply]
> this is the file on my .local file now (from my tests I believe it ADDs to the same value in okular.profile)

Correct, private-bin options accumulate (and support gmobbing).

> ...I guess we can add this on a common-kde.inc file? Is there such file?

We don't have such a file. And even if we did, it would take implementing a new conditional like `?HAS_PRIVATE_BIN: private-bin foo`. Until someone implements all of this the best you can do is add `private-bin ps2pdf` to any profile you want/need to have printing support. Suboptimal, I do realize that, but such is the current situation I'm afraid.

> the okular.profile one have things which I believe are very out of date

Thanks for sharing this info. Although I'm not aware of any written statement of how to deal with `outdated` parts of a profile, the `working assumption` is to not drop anything until there's proof it breaks a sandbox...
 
All in all a very nice piece of detective work to find that missing piece :) Can you open a PR to add this to the okular.profile at least?

**Reply by gcb** [reply]
Just one more complication... i noticed i can print to pdf from `kate` without allowing `ps2pdf` in any way i can detect... sigh.

**Reply by ghost** [reply]
> Just one more complication... i noticed i can print to pdf from kate without allowing ps2pdf in any way i can detect... sigh.

That's because kate.profile `doesn't` have `private-bin`, so the needed binaries are already accessible in its sandbox. Nothing to be done here.

**Reply by gcb** [reply]
nono, i was blocking the binaries on kate and it still worked. I asked about on the okular irc, and it seems it is still not fully utilizing the pure-QT solution: https://invent.kde.org/graphics/okular/-/merge_requests/411

I believe most KDE/QT things up to date should not require access to external bin, besides the QT libs.

**Comment by kmk3** [comment, marked as answer]
Should be fixed by:

* #6408
